apm-lambda-extension/extension/apm_server.go

```diff
@@ -72,19 +106,79 @@ func PostToApmServer(client *http.Client, agentData AgentData, config *extension
 		req.Header.Add("Authorization", "Bearer "+config.apmServerSecretToken)
 	}
 
+	Log.Debug("Sending data chunk to APM Server")
 	resp, err := client.Do(req)
 	if err != nil {
+		enterBackoff(ctx)
 		return fmt.Errorf("failed to post to APM server: %v", err)
 	}
 
 	//Read the response body
 	defer resp.Body.Close()
 	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
+		enterBackoff(ctx)
 		return fmt.Errorf("failed to read the response body after posting to the APM server")
 	}
 
+	apmServerTransportState.Status = Healthy
+	Log.Debug("Transport status set to healthy")
 	Log.Debugf("APM server response body: %v", string(body))
 	Log.Debugf("APM server response status code: %v", resp.StatusCode)
 	return nil
 }
+
+func EnqueueAPMData(agentDataChannel chan AgentData, agentData AgentData) {
+	select {
+	case agentDataChannel <- agentData:
+		Log.Debug("Adding agent data to buffer to be sent to apm server")
+	default:
+		Log.Warn("Channel full: dropping a subset of agent data")
+	}
+}
+
+func IsTransportStatusHealthyOrPending() bool {
+	return apmServerTransportState.Status != Failing
+}
+
+func WaitForGracePeriod(ctx context.Context) {
+	select {
+	case <-apmServerTransportState.GracePeriodTimer.C:
+		Log.Debug("Grace period over - timer timed out")
+		return
+	case <-ctx.Done():
+		Log.Debug("Grace period over - context done")
+		return
+	}
+
+}
+
+// Warning : the apmServerTransportStatus state needs to be locked if this function is ever called
+// concurrently in the future.
+func enterBackoff(ctx context.Context) {
+	Log.Info("Entering backoff")
+	if apmServerTransportState.Status == Healthy {
+		apmServerTransportState.ReconnectionCount = 0
+		Log.Debug("Entered backoff as healthy : reconnection count set to 0")
+	} else {
+		apmServerTransportState.ReconnectionCount++
+		Log.Debugf("Entered backoff as pending : reconnection count set to %d", apmServerTransportState.ReconnectionCount)
+	}
+	apmServerTransportState.Status = Failing
+	Log.Debug("Transport status set to failing")
+	apmServerTransportState.GracePeriodTimer = time.NewTimer(computeGracePeriod())
+	go func() {
+		apmServerTransportState.Lock()
+		defer apmServerTransportState.Unlock()
+		WaitForGracePeriod(ctx)
+		apmServerTransportState.Status = Pending
+		Log.Debug("Transport status set to pending")
+	}()
```

Comment:
Added locks in the embedded goroutine. I believe this makes sense from a design perspective (lock the transport state in `Failing` while we wait for the grace period timer to be elapsed). Such locks are also required for the tests in `apm_server_test.go` to pass : see dedicated comment.